lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,368 @@
+            return -1;    // Bad token
+        }
+
+        if (verbose) {
+            std::cout << std::setw(20) << token << " ";
+            for (int i = 0; i < value.size(); i++) std::cout << value[i] << " ";
+            std::cout << std::endl;
+        }
+        return 1;    // we now have a token
+    }
+    return 0;    // we didn't find the token
+}
+
+void BOVCollection::_findTokenValue(std::string &line) const
+{
+    std::string delimiter = ": ";
+
+    size_t      pos = 0;
+    std::string token;
+    while ((pos = line.find(delimiter)) != std::string::npos) {
+        token = line.substr(0, pos);
+        line.erase(0, pos + delimiter.length());
+    }
+}
+
+size_t BOVCollection::_sizeOfFormat(DC::XType type) const
+{
+    switch (type) {
+    case DC::XType::INT32: return 4;
+    case DC::XType::FLOAT: return 4;
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    FILE *fp = fopen(_dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %d", fp);
+        return -1;
+    }
+
+    size_t formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        return (-1);
+    }
+
+    if (_dataSize.size() != 3) {
+        SetErrMsg("Invalid grid size (must be 3D)");
+        return (-1);
+    }
+    size_t numValues = _dataSize[0] * _dataSize[1] * _dataSize[2];
+
+    int  n = 1;
+    bool systemLittleEndian = *(char *)&n == 1 ? true : false;
+    bool dataLittleEndian = _dataEndian == "LITTLE" ? true : false;
+    bool needSwap = systemLittleEndian != dataLittleEndian ? true : false;
+
+    // Read a "pencil" of data along the X axis, one row at a time
+    size_t count = max[0] - min[0] + 1;
+    for (int k = min[2]; k <= max[2]; k++) {
+        int zOffset = _dataSize[0] * _dataSize[1] * k;
+        for (int j = min[1]; j <= max[1]; j++) {
+            int xOffset = min[0];
+            int yOffset = _dataSize[0] * j;
+            int offset = formatSize * (xOffset + yOffset + zOffset);
+
+            unsigned char readBuffer[count * formatSize];
+
+            if (needSwap) { _swapBytes(readBuffer, formatSize, numValues); }
+
+            fseek(fp, offset, SEEK_SET);
+            size_t rc = fread(readBuffer, formatSize, count, fp);
+
+            if (rc != count) {
+                if (ferror(fp) != 0) {
+                    MyBase::SetErrMsg("Error reading input file");
```

Comment:
For C library calls that set errno use "%M" in SetErrMsg to get the error message associated with errno:

```suggestion
                    MyBase::SetErrMsg("Error reading input file: %M");
```

Reply:
Didn't know that.  Fixed.